File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int unused = 0; unused < n+1; unused ++)
            graph.add(new ArrayList<>());
        
        for (int[] wire: wires) {
            int a = wire[0];
            int b = wire[1];
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        int[] weight = new int[n+1];
        dfs(graph, weight, 1, 0);
        
        int mn = n;
        for (int i = 1; i <= n; i++)
            mn = Math.min(mn, Math.abs(n - 2 * weight[i]));
        
        return mn;
    }
    
    public int dfs(List<List<Integer>> graph, int[] weight, int now, int bef) {
        int ret = 1;
        for (int nxt: graph.get(now)) {
            if (bef == nxt)
                continue;
            
            ret += dfs(graph, weight, nxt, now);
        }
        weight[now] = ret;
        return ret;
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    graph = [[] for _ in range(n+1)]
    for u, v in wires:
        graph[u].append(v)
        graph[v].append(u)
        
    counts = [0] * (n+1)
    dfs(graph, counts, 1, 0)
    
    return min(map(lambda idx: abs(n-2*counts[idx]), range(1, n+1)))
        
        
def dfs(graph, counts, now, bef):
    ret = 1
    for nxt in graph[now]:
        if nxt == bef:
            continue
            
        ret += dfs(graph, counts, nxt, now)
        
    counts[now] = ret;
    return ret
        